### ISSUE TITLE
DX12 Vertex Shader - Pixel Shader linkage error

### ISSUE THREAD

**Issue by LPGhatguy** [gfx-rs/wgpu#1171]
**Description**
When running my application using the DX12 backend, I receive an error from `gfx_backend_dx12` along with a D3D12 error when I have a debugger attached:

```
D3D12 ERROR: ID3D12Device::CreateGraphicsPipelineState: Vertex Shader - Pixel Shader linkage error:
Signatures between stages are incompatible. Semantic 'TEXCOORD' is defined for mismatched hardware
registers between the output stage and input stage.
[ STATE_CREATION ERROR #660: CREATEGRAPHICSPIPELINESTATE_SHADER_LINKAGE_REGISTERINDEX]

[ERROR gfx_backend_dx12::device] Failed to build shader: 80070057
thread 'main' panicked at 'failed to create graphics pipeline: Implementation specific error occurred',
C:\Users\lpgla\projects\wgpu\wgpu-core\src\device\mod.rs:2267:26
```

This project runs without validation errors on using the Vulkan backend.

Here are the shaders I'm using: https://gist.github.com/LPGhatguy/b04ceb08f367ed46baf4b1fa8d111af6

Here is the full program output using `RUST_LOG=gfx_backend_dx12=debug,info`: https://gist.github.com/LPGhatguy/2a233e225ab0ec1e00abe1ad3d0c2cf4

**Repro steps**
The project this occurs in is not yet public and I haven't reduced the case. I'm hoping the given logs are enough to help; if not, let me know!

**Platform**
Windows 10 v1909, wgpu c2c8d2203ed312e3b63354096ec90634b8f9b0b5

Shaders compiled with this glslc info:
```
$ glslc --version
shaderc v2020.4-dev v2020.3-11-g702723a
spirv-tools v2020.7-dev v2020.6
glslang 10-11.0.0-23-gdd69df7f

Target: SPIR-V 1.0
```

**Comment by kvark** [comment]
Interesting stuff! Looking at the shaders, it's not obvious to me what it doesn't like. Maybe it's the `float3x3 tbn` stuff?
Would you be able to try splitting this over to 3 different vectors, just for a test?

Another thing - you could provide the full reproducible test case without opening the source code by recording an API trace. See https://github.com/gfx-rs/wgpu/wiki/Debugging-wgpu-Applications#tracing-infrastructure. If you do this, just zip it and drop in here. We'd be able to investigate further. Note: of course, messing with compiled SPIR-V shaders is not as convenient here as for you modifying the GLSL sources, hence I asked if you are able to try this change first.

**Comment by LPGhatguy** [comment]
I tried breaking apart the `tbn` matrix into three vectors and it seems like I got the same result. Changed the vertex shader output to instead read as

```glsl
layout(location = 0) out VS_OUT {
    vec3 pos;
    vec2 tex_coord;
    vec3 t;
    vec3 b;
    vec3 n;
} vs_out;
```

along with the matching fragment changes.

I also recorded a trace with the issue, attached here: [cyrtanthus-dx12-trace.zip](https://github.com/gfx-rs/wgpu/files/5871121/cyrtanthus-dx12-trace.zip)

**Comment by kvark** [comment]
@tangmi discovered that `SV_Position` output declared first is shifting all the internal registers by one in FXC.
So this appears to be a bug in FXC that could be worked around at SPIRV-Cross or Naga level, but is currently not.

You could try messing with the GLSL code to either move the `gl_Position` use after your varyings, or make the fragment shader use something like `gl_FragCoord` to also occupy this first slot...

**Comment by tangmi** [comment]
Yeah, the error message seems like it could be related to the register assigned to `SV_Position`.

Given the output generated shaders from https://gist.github.com/LPGhatguy/2a233e225ab0ec1e00abe1ad3d0c2cf4#file-output-txt-L296-L501 and running them through `fxc` (`dxc` does similar things with shader model 6.x):

`fxc /T vs_5_1 .\a.vs.hlsl` makes the following output signature:
```
// Output signature:
//
// Name                 Index   Mask Register SysValue  Format   Used
// -------------------- ----- ------ -------- -------- ------- ------
// SV_Position              0   xyzw        0      POS   float   xyzw
// TEXCOORD                 0   xyz         1     NONE   float   xyz
// TEXCOORD                 1   xy          2     NONE   float   xy
// TEXCOORD                 2   xyz         3     NONE   float   xyz
// TEXCOORD                 3   xyz         4     NONE   float   xyz
// TEXCOORD                 4   xyz         5     NONE   float   xyz
```

and `fxc /T ps_5_1 .\a.ps.hlsl` makes the following input signature
```
// Input signature:
//
// Name                 Index   Mask Register SysValue  Format   Used
// -------------------- ----- ------ -------- -------- ------- ------
// TEXCOORD                 0   xyz         0     NONE   float   xyz
// TEXCOORD                 1   xy          1     NONE   float   xy
// TEXCOORD                 2   xyz         2     NONE   float   xyz
// TEXCOORD                 3   xyz         3     NONE   float   xyz
// TEXCOORD                 4   xyz         4     NONE   float   xyz
```

It seems like SPIRV-Cross automatically outputs a `SPIRV_Cross_Output` type (containing the `SV_Position` semantic) for you as a return value, which I guess `fxc` and `dxc` will put above `out` params when assigning registers.

Here's some related posts I've found around the web:

- https://stackoverflow.com/a/37740051 <- this one seems to imply this is some sketchy behavior that varies across AMD/NVidia
- https://gamedev.net/forums/topic/665069-vertex-shader-pixel-shader-linkage-error-signatures-between-stages-are-incompatible/5206119/

I was able to modify the SPIRV-Cross HLSL generated output to include `float4  gl_Position : SV_Position` at the end of `VSOut` to cause `fxc` to make all the `TEXCOORD`s line up register-wise--I'm unsure if this actually fixes your problem, however...

**Comment by kvark** [comment]
Recent [discord discussion](https://discord.com/channels/590611987420020747/590965930658496690) on this.
TL;DR: looks like an FXC bug that was acknowledged internally but hasn't fixed since 2017.
For the workaround, we need to declare `SV_Position` as the first input of pixel shaders. I think that should be sufficient.

**Comment by seanbaxter** [comment]
Declaring SV_Position as the first input won't fix anything. This is a really challenging D3D bug if your frontend doesn't strictly order interface variables, like how HLSL orders them by making them function parameters. Additionally, if your vertex shader uses/declares output variables, all of those must be accounted for as inputs in the pixel shader in order to get the mapping correct.

This is going to be an existential problem for frontends that use the GLSL-like convention of declaring interface variables as globals and then including them in the EntryPoint when ODR-used.

**Comment by kvark** [comment]
Pretty sure we solved this now
